platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java

```diff
@@ -990,13 +1019,68 @@ public Stream<VirtualHashRecord> dirtyHashesForFlush(final long lastLeafPath) {
     }
 
     /**
-     * Gets estimated number of dirty internal nodes in this cache.
+     * Gets estimated number of dirty hashes in this cache.
+     *
+     * <p>This method and similar methods about leaves are called multiple times. First,
+     * the method is used to estimate this cache copy size to make a decision whether to
+     * flush the copy to disk or to merge it to a newer copy. If the copy is selected to
+     * flush, its {@link #prepareForFlush()} method is called, it may mark some mutations
+     * as filtered, which affects the estimated size. After that this method is called
+     * again, it takes filtered mutations into consideration. If the updated estimated
+     * size is lower than flush threshold, the copy is not flushed to disk, but rather
+     * its {@link #garbageCollect(long, long)} is called.
      *
      * @return
-     *        Estimated number of dirty internal nodes
+     *        Estimated number of dirty hashes
+     */
+    public long estimatedHashesCount() {
+        return (dirtyHashes == null) ? 0 : (dirtyHashes.size() - filteredHashesCount.get());
+    }
+
+    /**
+     * This cache copy is selected to flush to disk, either because it is explicitly
+     * marked as such, or when cache copy estimated size exceeds flush threshold. Before
+     * flush, this method is used to mark redundant mutations as "filtered", so they aren't
+     * included to the streams for the data source. It may happen that after filtering the
+     * cache copy is no longer needed to flush. In this case, all filtered mutations are
+     * removed using {@link #garbageCollect(long, long)} method.
+     */
+    public void prepareForFlush() {
+        // Mark obsolete mutations to filter later and update "filtered" counters. These
+        // counters will affect the estimated size
+        final long version = getFastCopyVersion();
+        filterMutations(dirtyHashes, pathToDirtyHashIndex, version, filteredHashesCount);
+        filterMutations(dirtyLeafPaths, pathToDirtyLeafIndex, version, filteredLeafPathsCount);
+        filterMutations(dirtyLeaves, keyToDirtyLeafIndex, version, filteredLeavesCount);
+    }
+
+    /**
+     * If a cache copy's estimated size exceeded flush threshold, but after mutations are
+     * filtered the size drops below the threshold, the copy is not flushed to disk, but
+     * just removes all filtered mutations using this method.
      */
-    public long estimatedInternalsCount(final long firstLeafPath) {
-        return (dirtyHashes == null) ? 0 : dirtyHashes.size();
+    public void garbageCollect(final long firstLeafPath, final long lastLeafPath) {
+        final Stream<Mutation<Long, Hash>> filteredHashes = dirtyHashes.stream()
+                .filter(mutation -> mutation.key <= lastLeafPath)
+                .filter(mutation -> !mutation.isFiltered())
+                .peek(mutation -> mutation.next = null);
```

Comment:
The consequences of this operation is unclear. From the look of it, some sort of a linked list gets cut. However, original objects remain in the original `dirtyHashes` collection. Also, assuming that the original linked list is not circular, it's unclear how the object with `next` cleared here would differ from the actual tail of the original list. I suggest to add a clarifying comment in code to explain what exactly the code is trying to achieve here and where/how the (potentially) nullified `next` value is going to be used in other places.

Same comment applies to a couple of very similar streams below in this same method.

Reply:
It's slightly more complicated than just a linked list... All cache instances share a concurrent map `pathToDirtyHashIndex`, which is a map of paths (longs) to linked lists (hash mutations). In addition to this map, every cache instance contains an array `dirtyHashes`, array elements are individual mutations from the linked lists above. Each mutation in the array points to the next mutation of the same hash, but not to the next element in the array. By setting `mutation.next` to null in this code, the linked lists in the map get trimmed, all elements after certain mutations are removed.

I added a short comment to this `peek()` call:

```java
                // Leave only the latest mutation for every hash path by setting next to null
```